### PR TITLE
Lazily call preprocess to shorten vcgen time

### PR DESCRIPTION
**Pull request by aqjune**
This is a patch to lazily call preprocess to shorten VCGen time.

### CODE REVIEW

File: tools/transform.cpp [hunk 1]
```diff
@@ -436,15 +436,15 @@ check_refinement(Errors &errs, Transform &t, State &src_state, State &tgt_state,
     return;
   }
 
-  auto mk_fml = [&](expr &&refines) -> expr {
+  auto mk_fml = [&](const expr &refines) -> expr {
```
Comment (aqjune) [inline tools/transform.cpp:439]: std::bind wasn't working well for lambda functions that have arguments with rvalue reference.
The reason is that std::bind internally forwards the argument to a variable and calls the function: https://stackoverflow.com/a/34878230

**Comment by nunoplopes** [comment]
This patch made me think that we need to shake things up in that function.
We produce all formulas upfront, which increase peak memory consumption. It would be best if we produced the formulas only for the query and then destroy them straight away to reclaim the memory.

**Comment by aqjune** [comment]
That really makes sense..! I'll work on this.

**Comment by aqjune** [comment]
I'll see whether there is no regression in single file benchmarks.
Since the diff is slightly big, I'm fine with having it reviewed after the submission.

**Comment by aqjune** [comment]
No regression in bzip2/gzip/oggenc/ph7 (sqlite3 is running)

**Comment by aqjune** [comment]
no regression in sqlite3 too

**Pull request merged**